[reciever/prometheusremotewritereciever] handle otel_scope_name and otel_scope_version in prw reciever

Vandit1604 opened the pull request:
This PR ensures that when Prometheus samples lack the otel_scope_name or otel_scope_version labels, the receiver fills in default values. If otel_scope_name is missing, it defaults to `opentelemetry-collector`. If otel_scope_version is missing, it uses the collector’s build version. This change makes sure every metric has a complete instrumentation scope as required.

more here: [Metrics which do not have an otel_scope_name or otel_scope_version label MUST be assigned an instrumentation scope identifying the entity performing the translation from Prometheus to OpenTelemetry](https://opentelemetry.io/docs/specs/otel/compatibility/prometheus_and_openmetrics/#instrumentation-scope)

Vandit1604 committed:
handle otel_scope_name and otel_scope_version in prw reciever

Signed-off-by: Vandit Singh <vanditsinghkv@gmail.com>

Vandit1604 commented:
Should I also handle `otel_scope_info` in this?

> Labels on otel_scope_info metric points other than otel_scope_name and otel_scope_version, MUST be added as scope attributes to the scope with the matching name and version.

found this in the OTLP spec [here](https://opentelemetry.io/docs/specs/otel/compatibility/prometheus_and_openmetrics/#instrumentation-scope)

dashpole commented:
> Should I also handle otel_scope_info in this?

Not right now.  That spec needs to be updated because scope attributes were made identifying

ArthurSens commented on receiver/prometheusremotewritereceiver/receiver.go:
```diff
@@ -230,15 +234,24 @@ func addCounterDatapoints(_ pmetric.ResourceMetrics, _ labels.Labels, _ writev2.
+	// If the scope version or scope name is empty, get the information from the collector build tags.
 	// More: https://opentelemetry.io/docs/specs/otel/compatibility/prometheus_and_openmetrics/#:~:text=Metrics%20which%20do%20not%20have%20an%20otel_scope_name%20or%20otel_scope_version%20label%20MUST%20be%20assigned%20an%20instrumentation%20scope%20identifying%20the%20entity%20performing%20the%20translation%20from%20Prometheus%20to%20OpenTelemetry%20(e.g.%20the%20collector%E2%80%99s%20prometheus%2 ...
+	if scopeName == "" {
+		scopeName = "opentelemetry-collector"
```
Can't we take the scope name from `settings,` just like you did with the version?

Vandit1604 replied:
something like this ?

``` 
scopeName = prw.settings.BuildInfo.Command
```

sonalgaud12 replied:
Um, how about adding scopedetail func? Which could fetch data from collector setting

ArthurSens replied:
I think Vandit needs a bit more detail about what a scopedetails func is. Looking at your PR https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/37871, I like the idea, but let's explain why this is good. :)

This is also unrelated to this specific thread, so let's start a new review for this idea

ArthurSens replied:
> something like this ?
> 
> ```
> scopeName = prw.settings.BuildInfo.Command
> ```

When reading the doc string for `settings.BuildInfo.Description`, I see:

```
Description is the full name of the collector, e.g. "OpenTelemetry Collector"
```

So I think the description is the most appropriate one.

sonalgaud12 replied:
Understood thanks :)

ArthurSens commented on receiver/prometheusremotewritereceiver/receiver.go:
```diff
@@ -230,15 +234,24 @@ func addCounterDatapoints(_ pmetric.ResourceMetrics, _ labels.Labels, _ writev2.
 	// TODO: Implement this function
 }
 
+var defaultBuildVersion string = "unknown"
```
I'm not sure if we want to do this. Have you seen this requirement somewhere?

Vandit1604 replied:
Not specifically, I added this as a fallback in case `settings.BuildInfo.Version` is empty or unavailable. Is this not the correct approach? Happy to adjust if there's a better way to handle it.

PiyushRajDev replied:
I noticed the test failures seem related to scope handling. It looks like the tests expect empty scopes in some cases, but our changes always set "opentelemetry-collector" as the default scope.

Should we:
Update the tests to expect this new default scope, or
Only set the default scope in specific conditions?
Looking at receiver_test.go:258, the test cases seem to verify behavior when scopes are missing. Happy to help adjust either the implementation or the tests based on what’s preferred

ArthurSens reviewed (commented):
Thanks for starting the work here :) 

We'll also need tests to verify that we're not breaking anything. Can you add one test to ensure that two time series without scope labels end up together and with the correct scope attributes?

ArthurSens reviewed (commented):
I think we're over engineering a bit with the fallbacks, there's nothing in the spec asking for this so we can just use empty strings.

Vandit1604 committed:
remove the fallbacks and use description as otel_scope_name

Signed-off-by: Vandit Singh <vanditsinghkv@gmail.com>

Vandit1604 committed:
fix up test

Signed-off-by: Vandit Singh <vanditsinghkv@gmail.com>

Vandit1604 committed:
add tests

Signed-off-by: Vandit Singh <vanditsinghkv@gmail.com>

Vandit1604 commented:
Hi @ArthurSens I have added tests and changelog entry as well.

Please review in your own time.
Thank you!

perebaj commented on receiver/prometheusremotewritereceiver/receiver_test.go:
```diff
@@ -219,33 +219,123 @@ func TestTranslateV2(t *testing.T) {
+			expectedStats: remote.WriteResponseStats{},
+		},
+		{
+			name: "missing otel_scope_name/version falls back to buildName/buildVersion",
```
Both `missing otel_scope_name/version falls back to buildName/buildVersion` and `valid request` are testing the same scenario, no? In both, we are not passing otel_scope_name/version and expecting that the current value will be the one that comes from buildName/Version. Maybe we can create a test case where the we assign an value to the settings.BuildInfo.Description/Version and validate here, if the value was properly assigned.

perebaj commented on receiver/prometheusremotewritereceiver/receiver.go:
```diff
@@ -28,6 +28,14 @@ import (
 )
 
 func newRemoteWriteReceiver(settings receiver.Settings, cfg *Config, nextConsumer consumer.Metrics) (receiver.Metrics, error) {
+	if settings.BuildInfo.Version != "" {
+		buildVersion = settings.BuildInfo.Version
+	}
+
+	if settings.BuildInfo.Description != "" {
+		buildName = settings.BuildInfo.Description
+	}
+
 	return &prometheusRemoteWriteReceiver{
```
Regardless of whether `settings.BuildInfo.Version` or `settings.BuildInfo.Description` are empty, we should always assign their values to buildName and version. The determining factor for whether we use them or not will be whether scope_name or version are empty. I think that we can simplify the `ifs` here.

```suggestion
	buildVersion = settings.BuildInfo.Version

    buildName = settings.BuildInfo.Description

	return &prometheusRemoteWriteReceiver{
```

ArthurSens replied:
True :)

We don't even need the extra variables, actually. We can update the function like this:

```go
func (prwr * prometheusRemoteWriteReceiver) addGaugeDatapoints(..... {
  scopeName := prwr.settings.BuildInfo.Description
  scopeVersion := prwr.settings.BuildInfo.Version
}
```

If the settings are empty, the scope variables will be empty. If settings had proper content, then the scope will be populated correctly

perebaj commented on receiver/prometheusremotewritereceiver/receiver.go:
```diff
@@ -230,15 +238,28 @@ func addCounterDatapoints(_ pmetric.ResourceMetrics, _ labels.Labels, _ writev2.
 	// TODO: Implement this function
 }
 
+var (
+	buildVersion string = ""
+	buildName    string = ""
+)
```
According to [this comment](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/37791#discussion_r1953584912), we don't even need to set those vars. So, ignore my suggestion 🙃

ArthurSens commented:
Could someone with permissions trigger tests here? Maybe @dashpole :)

Vandit1604 committed:
implement suggestions and update test

Signed-off-by: Vandit Singh <vanditsinghkv@gmail.com>

Vandit1604 commented:
Updated the test with the case `custom BuildInfo used when no scope provided` to address @perebaj 's comment. Also updated the code to only override with collector build details if `ls.Get()` returns an empty string.

perebaj commented on receiver/prometheusremotewritereceiver/receiver_test.go:
```diff
@@ -219,35 +234,149 @@ func TestTranslateV2(t *testing.T) {
+				sm1.Scope().SetName(defaultBuildName)
+				sm1.Scope().SetVersion(defaultBuildVersion)
+				// Expect 2 separate gauge metrics, one per timeseries.
+				m1 := sm1.Metrics().AppendEmpty().SetEmptyGauge()
+				dp1 := m1.DataPoints().AppendEmpty()
+				dp1.Attributes().PutStr("d", "e")
+				dp1.Attributes().PutStr("foo", "bar")
```
do you have any reason to change `sm1Attributes := sm1.Metrics().AppendEmpty().SetEmptyGauge().DataPoints().AppendEmpty().Attributes()` ->

m1 := sm1.Metrics().AppendEmpty().SetEmptyGauge()
dp1 := m1.DataPoints().AppendEmpty()
dp1.Attributes().PutStr(...)
dp1.Attributes().PutStr(...)

?

Seems that the previous implementation it's ok

perebaj commented on receiver/prometheusremotewritereceiver/receiver_test.go:
```diff
@@ -219,35 +234,149 @@ func TestTranslateV2(t *testing.T) {
+			// When missing, ls.Get returns "" so the defaults from BuildInfo are preserved.
+			request: &writev2.Request{
+				Symbols: []string{
+					"",                // index 0
+					"__name__",        // index 1
+					"metric_no_scope", // index 2
+					"job",             // index 3
+					"service-z/xyz",   // index 4
+					"instance",        // index 5
+					"inst-42",         // index 6
+					"d",               // index 7
+					"e",               // index 8
+					"foo",             // index 9
+					"bar",             // index 10
```
```suggestion
					"",                
					"__name__",
					"metric_no_scope",
					"job",
					"service-z/xyz",
					"instance",
					"inst-42",
					"d",
					"e",
					"foo",
					"bar",
```

perebaj commented on receiver/prometheusremotewritereceiver/receiver_test.go:
```diff
@@ -219,35 +234,149 @@ func TestTranslateV2(t *testing.T) {
+				parseJobAndInstance(rm.Resource().Attributes(), "service-z/xyz", "inst-42")
+				sm := rm.ScopeMetrics().AppendEmpty()
+				// Expect fallback to default BuildInfo.
+				sm.Scope().SetName(defaultBuildName)
+				sm.Scope().SetVersion(defaultBuildVersion)
```
Maybe set the name and version as empty strings to make explicit the value that will be assigned for them?
```suggestion
				sm.Scope().SetName("")
				sm.Scope().SetVersion("")
```

perebaj commented on receiver/prometheusremotewritereceiver/receiver_test.go:
```diff
@@ -149,41 +160,45 @@ func TestTranslateV2(t *testing.T) {
-				sm1 := rm1.ScopeMetrics().AppendEmpty()
+				rm := expected.ResourceMetrics().AppendEmpty()
+				parseJobAndInstance(rm.Resource().Attributes(), "service-x/test", "107cn001")
+				// Scope "scope1" for first two timeseries.
```
```suggestion
```

perebaj commented on receiver/prometheusremotewritereceiver/receiver_test.go:
```diff
@@ -149,41 +160,45 @@ func TestTranslateV2(t *testing.T) {
+				m2 := sm1.Metrics().AppendEmpty().SetEmptyGauge()
+				dp2 := m2.DataPoints().AppendEmpty()
+				dp2.Attributes().PutStr("d", "e")
+				// Scope "scope2" for the third timeseries.
```
Same
```suggestion
```

perebaj commented on receiver/prometheusremotewritereceiver/receiver_test.go:
```diff
@@ -121,15 +126,21 @@ func TestHandlePRWContentTypeNegotiation(t *testing.T) {
 
 func TestTranslateV2(t *testing.T) {
 	prwReceiver := setupMetricsReceiver(t)
+	// Save the default BuildInfo values.
```
```suggestion
```

perebaj reviewed (commented):
Some suggestions ✌🏽

Vandit1604 committed:
Apply suggestions from code review

Co-authored-by: Jonathan  <perebaj@gmail.com>

Vandit1604 committed:
fix failing tests

Signed-off-by: Vandit Singh <vanditsinghkv@gmail.com>

perebaj commented on receiver/prometheusremotewritereceiver/receiver_test.go:
```diff
@@ -323,8 +321,8 @@ func TestTranslateV2(t *testing.T) {
-				sm.Scope().SetName("")
-				sm.Scope().SetVersion("")
+				sm.Scope().SetName(defaultBuildName)
+				sm.Scope().SetVersion(defaultBuildVersion)
```
Using the empty string instead of defaultBuildName/Version does the test fail?

perebaj replied:
Hey @Vandit1604 I think that is only thing that is missing before we merge it 🙃

perebaj replied:
And some conflicts related to [this PR](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/37895/files#diff-e26ac1e595473c2a7f1cf349aee9b1cceced9c8f9c4615b08be8b9e551edfcd9) that was merged recently

ArthurSens commented on receiver/prometheusremotewritereceiver/receiver_test.go:
```diff
@@ -219,35 +231,147 @@ func TestTranslateV2(t *testing.T) {
 				sm1Attributes := sm1.Metrics().AppendEmpty().SetEmptyGauge().DataPoints().AppendEmpty().Attributes()
 				sm1Attributes.PutStr("d", "e")
 				sm1Attributes.PutStr("foo", "bar")
-				// Since we don't check "scope_name" and "scope_version", we end up with duplicated scope metrics for repeated series.
-				// TODO: Properly handle scope metrics.
```
nice, thanks for removing that :)

ArthurSens commented on receiver/prometheusremotewritereceiver/receiver_test.go:
```diff
@@ -219,35 +231,147 @@ func TestTranslateV2(t *testing.T) {
-				rmAttributes1.PutStr("service.instance.id", "107cn001")
+				parseJobAndInstance(rm1.Resource().Attributes(), "service-x/test", "107cn001")
 				sm1 := rm1.ScopeMetrics().AppendEmpty()
+				sm1.Scope().SetName(defaultBuildName)
+				sm1.Scope().SetVersion(defaultBuildVersion)
```
We don't need a fallback, let's keep an empty string

ArthurSens commented on receiver/prometheusremotewritereceiver/receiver_test.go:
```diff
@@ -219,35 +231,147 @@ func TestTranslateV2(t *testing.T) {
 				sm2Attributes.PutStr("d", "e")
 				sm2Attributes.PutStr("foo", "bar")
 
+				// For job "foo" with instance "bar", fallback to BuildInfo.
```
No need for a fallback

ArthurSens commented on receiver/prometheusremotewritereceiver/receiver_test.go:
```diff
@@ -219,35 +231,147 @@ func TestTranslateV2(t *testing.T) {
+			expectedStats: remote.WriteResponseStats{},
+		},
+		{
+			name: "provided otel_scope_name and otel_scope_version",
```
I'm not sure if I'm missing something, but I think this test case is covered by the test case `duplicated scope name and version`, isn't it? 

If that correct, could we remove this test case?

ArthurSens commented on receiver/prometheusremotewritereceiver/receiver_test.go:
```diff
@@ -219,35 +231,147 @@ func TestTranslateV2(t *testing.T) {
+			expectedStats: remote.WriteResponseStats{},
+		},
+		{
+			name: "missing otel_scope_name/version falls back to BuildInfo",
```
We don't need a fallback :) 

But we could certify that all metrics with a missing scope would end up within the same scope with no attributes.

ArthurSens commented on receiver/prometheusremotewritereceiver/receiver_test.go:
```diff
@@ -219,35 +231,147 @@ func TestTranslateV2(t *testing.T) {
+		{
+			name: "custom BuildInfo used when no scope provided",
+			// Even if BuildInfo is overridden, if no otel_scope is provided, the overridden defaults are used.
+			buildNameOverride:    "customBuildName",
+			buildVersionOverride: "customBuildVersion",
```
I don't really understand where this is coming from 🤔 could you explain?

ArthurSens commented on receiver/prometheusremotewritereceiver/receiver_test.go:
```diff
@@ -121,15 +126,20 @@ func TestHandlePRWContentTypeNegotiation(t *testing.T) {
+		expectError          string
+		expectedMetrics      pmetric.Metrics
+		expectedStats        remote.WriteResponseStats
+		buildNameOverride    string
+		buildVersionOverride string
```
What is this being used for? I don't really understand

ArthurSens reviewed (commented):
Thank you for being so diligent with the tests, I love it!

One extra comment besides the ones below: Could we keep all the scope-related tests together? What I mean is to move the test `duplicated scope name and version` further below, after `valid request`

Vandit1604 committed:
rebase and simplify tests

Signed-off-by: Vandit Singh <vanditsinghkv@gmail.com>

Vandit1604 committed:
methods should belong to prw

Signed-off-by: Vandit Singh <vanditsinghkv@gmail.com>

Vandit1604 committed:
group otel scope related tests

Signed-off-by: Vandit Singh <vanditsinghkv@gmail.com>

Vandit1604 commented:
I rebased the PR and did following changes
- Using settings.BuildInfo to get collector's build details
- The updated the tests accordingly.

perebaj commented:
Hey @Vandit1604 many thanks for your effort here, but this PR is a blocker for us. So we decided to go forward in a [simpler way](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/38413). But the others `addXDataPoints` should be implement using the `extractScopeInfo` approach

Vandit1604 commented:
@perebaj Should I close this PR then?
